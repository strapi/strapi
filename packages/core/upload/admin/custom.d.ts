export {};

declare global {
  interface Window {
    strapi: {
      backendURL: string;
    };
  }
  declare module '*?raw';
}
