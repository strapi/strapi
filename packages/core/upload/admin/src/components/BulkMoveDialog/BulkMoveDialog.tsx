import { Button, Flex, Grid, Field, Loader, Modal, Typography } from '@strapi/design-system';
import { Form, Formik, FormikErrors } from 'formik';
import isEmpty from 'lodash/isEmpty';
import { useIntl } from 'react-intl';

import { File } from '../../../../shared/contracts/files';
import { useBulkMove } from '../../hooks/useBulkMove';
import { useFolderStructure } from '../../hooks/useFolderStructure';
import { getTrad, normalizeAPIError } from '../../utils';
import { SelectTree } from '../SelectTree/SelectTree';

import type { Folder } from '../../../../shared/contracts/folders';
import type { OptionSelectTree } from '../SelectTree/SelectTree';
import type { FetchError } from '@strapi/admin/strapi-admin';

type InitialFormData = {
  destination:
    | {
        value: string | number;
        label: string;
      }
    | string;
};

interface FolderWithType extends Folder {
  type: string;
}

interface FileWithType extends File {
  type: string;
}

export interface BulkMoveDialogProps {
  onClose: () => void;
  selected?: Array<FolderWithType | FileWithType>;
  currentFolder?: FolderWithType;
}

export const BulkMoveDialog = ({ onClose, selected = [], currentFolder }: BulkMoveDialogProps) => {
  const { formatMessage } = useIntl();
  const { data: folderStructure, isLoading } = useFolderStructure();
  const { move } = useBulkMove();

  if (!folderStructure) {
    return null;
  }

  const handleSubmit = async (
    values: InitialFormData,
    { setErrors }: { setErrors: (errors: FormikErrors<InitialFormData>) => void }
  ) => {
    try {
      if (typeof values.destination !== 'string') {
        const destinationValue = values.destination.value;
        await move(destinationValue, selected);
        onClose();
      }
    } catch (error) {
      const normalizedError = normalizeAPIError(error as FetchError)!;

      if (normalizedError && 'errors' in normalizedError) {
        const formikErrors = normalizedError.errors?.reduce<Record<string, string>>(
          (acc, error) => {
            acc[error.values?.path?.length || 'destination'] = error.defaultMessage;

            return acc;
          },
          {}
        );

        if (!isEmpty(formikErrors)) {
          setErrors(formikErrors);
        }
      }
    }
  };

  if (isLoading) {
    return (
      <Modal.Content>
        <Modal.Body>
          <Flex justifyContent="center" paddingTop={4} paddingBottom={4}>
            <Loader>
              {formatMessage({
                id: getTrad('content.isLoading'),
                defaultMessage: 'Content is loading.',
              })}
            </Loader>
          </Flex>
        </Modal.Body>
      </Modal.Content>
    );
  }

  const initialFormData: InitialFormData = {
    destination: {
      value: currentFolder?.id || '',
      label: currentFolder?.name || folderStructure[0].label,
    },
  };

  return (
    <Modal.Content>
      <Formik validateOnChange={false} onSubmit={handleSubmit} initialValues={initialFormData}>
        {({ values, errors, setFieldValue }) => (
          <Form noValidate>
            <Modal.Header>
              <Modal.Title>
                {formatMessage({
                  id: getTrad('modal.folder.move.title'),
                  defaultMessage: 'Move elements to',
                })}
              </Modal.Title>
            </Modal.Header>

            <Modal.Body>
              <Grid.Root gap={4}>
                <Grid.Item xs={12} col={12} direction="column" alignItems="stretch">
                  <Field.Root id="folder-destination">
                    <Field.Label>
                      {formatMessage({
                        id: getTrad('form.input.label.folder-location'),
                        defaultMessage: 'Location',
                      })}
                    </Field.Label>

                    <SelectTree
                      options={folderStructure as OptionSelectTree[]}
                      onChange={(value: Record<string, string | number>) => {
                        setFieldValue('destination', value);
                      }}
                      defaultValue={
                        typeof values.destination !== 'string' ? values.destination : undefined
                      }
                      name="destination"
                      menuPortalTarget={document.querySelector('body')}
                      inputId="folder-destination"
                      error={errors?.destination}
                      ariaErrorMessage="destination-error"
                    />

                    {errors.destination && (
                      <Typography variant="pi" tag="p" textColor="danger600">
                        {errors.destination}
                      </Typography>
                    )}
                  </Field.Root>
                </Grid.Item>
              </Grid.Root>
            </Modal.Body>

            <Modal.Footer>
              <Modal.Close>
                <Button variant="tertiary" name="cancel">
                  {formatMessage({ id: 'cancel', defaultMessage: 'Cancel' })}
                </Button>
              </Modal.Close>
              <Button type="submit" loading={isLoading}>
                {formatMessage({ id: 'modal.folder.move.submit', defaultMessage: 'Move' })}
              </Button>
            </Modal.Footer>
          </Form>
        )}
      </Formik>
    </Modal.Content>
  );
};
