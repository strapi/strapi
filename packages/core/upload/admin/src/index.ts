import { Images } from '@strapi/icons';

import pluginPkg from '../../package.json';

import { MediaLibraryDialog } from './components/MediaLibraryDialog/MediaLibraryDialog';
import { MediaLibraryInput } from './components/MediaLibraryInput/MediaLibraryInput';
import { PERMISSIONS } from './constants';
import { pluginId } from './pluginId';
import { getTrad, prefixPluginTranslations } from './utils';

import type { MediaLibraryDialogProps } from './components/MediaLibraryDialog/MediaLibraryDialog';
import type { MediaLibraryInputProps } from './components/MediaLibraryInput/MediaLibraryInput';
import type { StrapiApp } from '@strapi/admin/strapi-admin';
import type { Plugin } from '@strapi/types';

const name = pluginPkg.strapi.name;

const admin: Plugin.Config.AdminInput = {
  register(app: StrapiApp) {
    app.addMenuLink({
      to: `plugins/${pluginId}`,
      icon: Images,
      intlLabel: {
        id: `${pluginId}.plugin.name`,
        defaultMessage: 'Media Library',
      },
      permissions: PERMISSIONS.main,
      Component: () => import('./pages/App/App').then((mod) => ({ default: mod.Upload })),
      position: 4,
    });

    app.addSettingsLink('global', {
      id: 'media-library-settings',
      to: 'media-library',
      intlLabel: {
        id: getTrad('plugin.name'),
        defaultMessage: 'Media Library',
      },
      async Component() {
        const { ProtectedSettingsPage } = await import('./pages/SettingsPage/SettingsPage');
        return { default: ProtectedSettingsPage };
      },
      permissions: PERMISSIONS.settings,
    });

    app.addFields({
      type: 'media',
      Component: MediaLibraryInput as React.FC<Partial<MediaLibraryInputProps>>,
    });
    app.addComponents([
      {
        name: 'media-library',
        Component: MediaLibraryDialog as React.FC<Partial<MediaLibraryDialogProps>>,
      },
    ]);

    app.registerPlugin({
      id: pluginId,
      name,
    });
  },
  async registerTrads({ locales }: { locales: string[] }) {
    const importedTrads = await Promise.all(
      locales.map((locale) => {
        return import(`./translations/${locale}.json`)
          .then(({ default: data }) => {
            return {
              data: prefixPluginTranslations(data, pluginId),
              locale,
            };
          })
          .catch(() => {
            return {
              data: {},
              locale,
            };
          });
      })
    );

    return Promise.resolve(importedTrads);
  },
};

// eslint-disable-next-line import/no-default-export
export default admin;
