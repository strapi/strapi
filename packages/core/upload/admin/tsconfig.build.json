{
  "extends": "tsconfig/client.json",
  "compilerOptions": {
    "rootDir": "../",
    "baseUrl": ".",
    "outDir": "../dist",
    "noEmit": false
  },
  "include": ["./src", "./custom.d.ts", "../shared", "../package.json"],
  "exclude": ["**/__mocks__", "./src/**/tests", "**/*.test.*"]
}
