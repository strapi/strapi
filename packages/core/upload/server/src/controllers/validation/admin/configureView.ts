import { yup, validateYupSchema } from '@strapi/utils';
import { ALLOWED_SORT_STRINGS } from '../../../constants';

const configSchema = yup.object({
  pageSize: yup.number().required(),
  sort: yup.mixed().oneOf(ALLOWED_SORT_STRINGS),
});

export const validateViewConfiguration = validateYupSchema(configSchema);

export type ViewConfiguration = yup.InferType<typeof configSchema>;
