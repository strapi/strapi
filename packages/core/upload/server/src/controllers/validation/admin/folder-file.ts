import { intersection, map, isEmpty } from 'lodash/fp';
import { yup, validateYupSchema } from '@strapi/utils';
import { FOLDER_MODEL_UID } from '../../../constants';
import { folderExists } from './utils';
import { isFolderOrChild } from '../../utils/folders';

import type { Folder } from '../../../types';

const validateDeleteManyFoldersFilesSchema = yup
  .object()
  .shape({
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateStructureMoveManyFoldersFilesSchema = yup
  .object()
  .shape({
    destinationFolderId: yup
      .strapiID()
      .nullable()
      .defined()
      .test('folder-exists', 'destination folder does not exist', folderExists),
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateDuplicatesMoveManyFoldersFilesSchema = yup
  .object()
  .test('are-folders-unique', 'some folders already exist', async function areFoldersUnique(value) {
    const { folderIds, destinationFolderId } = value;
    if (isEmpty(folderIds)) return true;

    const folders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { id: { $in: folderIds } },
    });

    const existingFolders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { parent: { id: destinationFolderId } },
    });

    const duplicatedNames = intersection(map('name', folders), map('name', existingFolders));
    if (duplicatedNames.length > 0) {
      return this.createError({
        message: `some folders already exists: ${duplicatedNames.join(', ')}`,
      });
    }

    return true;
  });

const validateMoveFoldersNotInsideThemselvesSchema = yup
  .object()
  .test(
    'dont-move-inside-self',
    'folders cannot be moved inside themselves or one of its children',
    async function validateMoveFoldersNotInsideThemselves(value) {
      const { folderIds, destinationFolderId } = value;
      if (destinationFolderId === null || isEmpty(folderIds)) return true;

      const destinationFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
        select: ['path'],
        where: { id: destinationFolderId },
      });

      const folders: Folder[] = await strapi.db.query(FOLDER_MODEL_UID).findMany({
        select: ['name', 'path'],
        where: { id: { $in: folderIds } },
      });

      const unmovableFoldersNames = folders
        .filter((folder) => isFolderOrChild(destinationFolder, folder))
        .map((f) => f.name);

      if (unmovableFoldersNames.length > 0) {
        return this.createError({
          message: `folders cannot be moved inside themselves or one of its children: ${unmovableFoldersNames.join(
            ', '
          )}`,
        });
      }

      return true;
    }
  );

export const validateDeleteManyFoldersFiles = validateYupSchema(
  validateDeleteManyFoldersFilesSchema
);

export async function validateMoveManyFoldersFiles(body: unknown) {
  await validateYupSchema(validateStructureMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateDuplicatesMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateMoveFoldersNotInsideThemselvesSchema)(body);
}
