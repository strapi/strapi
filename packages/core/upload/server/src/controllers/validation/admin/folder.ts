import { isUndefined, get, isNil } from 'lodash/fp';
import { yup, validateYupSchema } from '@strapi/utils';
import { getService } from '../../../utils';
import { FOLDER_MODEL_UID } from '../../../constants';
import { folderExists } from './utils';
import { isFolderOrChild } from '../../utils/folders';

const NO_SLASH_REGEX = /^[^/]+$/;
const NO_SPACES_AROUND = /^(?! ).+(?<! )$/;

const isNameUniqueInFolder = (id?: number): yup.TestFunction<string | undefined> => {
  return async function test(name?: string) {
    const { exists } = getService('folder');
    const filters: Record<string, unknown> = { name, parent: this.parent.parent || null };
    if (id) {
      filters.id = { $ne: id };

      if (isUndefined(name)) {
        const existingFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({ where: { id } });
        filters.name = get('name', existingFolder);
      }
    }

    const doesExist = await exists(filters);
    return !doesExist;
  };
};

const validateCreateFolderSchema = yup
  .object()
  .shape({
    name: yup
      .string()
      .min(1)
      .matches(NO_SLASH_REGEX, 'name cannot contain slashes')
      .matches(NO_SPACES_AROUND, 'name cannot start or end with a whitespace')
      .required()
      .test('is-folder-unique', 'A folder with this name already exists', isNameUniqueInFolder()),
    parent: yup
      .strapiID()
      .nullable()
      .test('folder-exists', 'parent folder does not exist', folderExists),
  })
  .noUnknown()
  .required();

const validateUpdateFolderSchema = (id: number) =>
  yup
    .object()
    .shape({
      name: yup
        .string()
        .min(1)
        .matches(NO_SLASH_REGEX, 'name cannot contain slashes')
        .matches(NO_SPACES_AROUND, 'name cannot start or end with a whitespace')
        .test(
          'is-folder-unique',
          'A folder with this name already exists',
          isNameUniqueInFolder(id)
        ),
      parent: yup
        .strapiID()
        .nullable()
        .test('folder-exists', 'parent folder does not exist', folderExists)
        .test(
          'dont-move-inside-self',
          'folder cannot be moved inside itself',
          async function test(parent) {
            if (isNil(parent)) return true;

            const destinationFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
              select: ['path'],
              where: { id: parent },
            });

            const currentFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
              select: ['path'],
              where: { id },
            });

            if (!destinationFolder || !currentFolder) return true;

            return !isFolderOrChild(destinationFolder, currentFolder);
          }
        ),
    })
    .noUnknown()
    .required();

export const validateCreateFolder = validateYupSchema(validateCreateFolderSchema);
export const validateUpdateFolder = (id: number) =>
  validateYupSchema(validateUpdateFolderSchema(id));
