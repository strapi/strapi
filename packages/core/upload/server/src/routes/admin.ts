export const routes = {
  type: 'admin',
  routes: [
    {
      method: 'GET',
      path: '/settings',
      handler: 'admin-settings.getSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/settings',
      handler: 'admin-settings.updateSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/',
      handler: 'admin-upload.upload',
      config: {
        policies: ['admin::isAuthenticatedAdmin'],
      },
    },
    {
      method: 'GET',
      path: '/files',
      handler: 'admin-file.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/files/:id',
      handler: 'admin-file.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'DELETE',
      path: '/files/:id',
      handler: 'admin-file.destroy',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders/:id',
      handler: 'admin-folder.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders',
      handler: 'admin-folder.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/folders',
      handler: 'admin-folder.create',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.create'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/folders/:id',
      handler: 'admin-folder.update',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folder-structure',
      handler: 'admin-folder.getStructure',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-delete',
      handler: 'admin-folder-file.deleteMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-move',
      handler: 'admin-folder-file.moveMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
  ],
};
