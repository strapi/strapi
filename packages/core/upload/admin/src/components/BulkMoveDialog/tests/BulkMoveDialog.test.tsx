import { NotificationsProvider } from '@strapi/admin/strapi-admin';
import { DesignSystemProvider, Modal } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { QueryClient, QueryClientProvider } from 'react-query';

import { BulkMoveDialog, BulkMoveDialogProps } from '../BulkMoveDialog';

jest.mock('../../../hooks/useFolderStructure');
jest.mock('../../../hooks/useBulkMove');

const client = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const ComponentFixture = (props: BulkMoveDialogProps) => {
  return (
    <QueryClientProvider client={client}>
      <IntlProvider locale="en" messages={{}}>
        <DesignSystemProvider>
          <NotificationsProvider>
            <Modal.Root open>
              <BulkMoveDialog {...props} />
            </Modal.Root>
          </NotificationsProvider>
        </DesignSystemProvider>
      </IntlProvider>
    </QueryClientProvider>
  );
};

function setup(props = { onClose: jest.fn(), selected: [] }) {
  return render(<ComponentFixture {...props} />, { container: document.getElementById('app')! });
}

describe('BulkMoveDialog', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders and matches the snapshot', () => {
    setup();
    expect(document.body).toMatchSnapshot();
  });
});
