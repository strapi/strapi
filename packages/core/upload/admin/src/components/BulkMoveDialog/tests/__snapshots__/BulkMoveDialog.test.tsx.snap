// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkMoveDialog renders and matches the snapshot 1`] = `
.c0 {
  margin-inline-start: -250px;
  top: 4.6rem;
  left: 50%;
  width: 50rem;
  z-index: 700;
  position: fixed;
}

.c5 {
  padding-block-start: 16px;
  padding-inline-end: 20px;
  padding-block-end: 16px;
  padding-inline-start: 20px;
  background: #f6f6f9;
}

.c9 {
  border-radius: 4px;
  display: inline-flex;
  cursor: pointer;
}

.c21 {
  padding-inline-end: 12px;
}

.c25 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  border-radius: 4px;
  cursor: pointer;
}

.c1 {
  gap: 8px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c6 {
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c10 {
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: flex;
}

.c16 {
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c18 {
  gap: 4px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c26 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c8 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: currentcolor;
  font-weight: 600;
}

.c19 {
  font-size: 1.2rem;
  line-height: 1.33;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: #32324d;
  font-weight: 600;
}

.c28 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c2 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c27 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
}

.c27:hover {
  background-color: #f6f6f9;
}

.c27:active {
  background-color: #eaeaef;
}

.c27[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c29 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #4945ff;
  background: #4945ff;
  color: #ffffff;
}

.c29:hover {
  border: 1px solid #7b79ff;
  background: #7b79ff;
}

.c29:active {
  border: 1px solid #4945ff;
  background: #4945ff;
}

.c29[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c11 {
  text-decoration: none;
  padding-block: 0.7rem;
  padding-inline: 0.7rem;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
  color: #8e8ea9;
}

.c11:hover {
  background-color: #f6f6f9;
  color: #666687;
}

.c11:active {
  background-color: #eaeaef;
}

.c11[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c20 {
  display: flex;
}

.c12 {
  width: 100%;
  height: 100%;
  overflow: hidden;
  display: flex;
}

.c14 {
  min-width: 100%;
}

.c3 {
  background: #32324d33;
  position: fixed;
  inset: 0;
  z-index: 300;
}

.c4 {
  max-width: 83rem;
  max-height: 90vh;
  height: auto;
  width: 60%;
  overflow: hidden;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  border-radius: 4px;
  background-color: #ffffff;
  box-shadow: 0px 2px 15px rgba(33, 33, 52, 0.1);
  z-index: 310;
}

.c7 {
  border-bottom: solid 1px #eaeaef;
}

.c13 {
  padding-inline: 40px;
}

.c13>div {
  padding-block: 40px;
}

.c13>div>div {
  display: block!important;
}

.c24 {
  border-top: solid 1px #eaeaef;
  flex: 1;
}

.c15 {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  gap: 16px;
}

.c17 {
  grid-column: span 12;
  max-width: 100%;
}

.c22 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c22 svg {
  height: 1.1rem;
  width: 1.1rem;
}

.c22 svg path {
  fill: #666687;
}

.c23 {
  display: flex;
  background: none;
  border: none;
}

.c23 svg {
  width: 0.9rem;
}

@media (prefers-reduced-motion: no-preference) {
  .c27 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c29 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c11 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c3 {
    animation: ececrQ 200ms cubic-bezier(.4,0,.2,1);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c4[data-state='open'] {
    animation-duration: 200ms;
    animation-timing-function: cubic-bezier(.4,0,.2,1);
    animation-name: iWCnzW;
  }

  .c4[data-state='closed'] {
    animation-duration: 120ms;
    animation-timing-function: cubic-bezier(0.25, 0.46, 0.45, 0.94);
    animation-name: jgufbq;
  }
}

@media (min-width: 520px) {
  .c17 {
    grid-column: span 12;
  }
}

@media (min-width: 768px) {
  .c17 {
    grid-column: span 12;
  }
}

@media (min-width: 1080px) {
  .c17 {
    grid-column: span 12;
  }
}

<body
  data-scroll-locked="1"
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <div
      aria-hidden="true"
      class="c0 c1"
      data-aria-hidden="true"
    />
    <span
      class="c2"
    >
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-log"
        role="log"
      />
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-status"
        role="status"
      />
      <p
        aria-live="assertive"
        aria-relevant="all"
        id="live-region-alert"
        role="alert"
      />
    </span>
  </div>
  <div
    class="c3"
    data-state="open"
    style="pointer-events: auto;"
  >
    <div
      aria-describedby="radix-:r2:"
      aria-labelledby="radix-:r1:"
      class="c4"
      data-state="open"
      id="radix-:r0:"
      role="dialog"
      style="pointer-events: auto;"
      tabindex="-1"
    >
      <form
        action="#"
        novalidate=""
      >
        <header
          class="c5 c6 c7"
        >
          <h2
            class="c8"
            id="radix-:r1:"
          >
            Move elements to
          </h2>
          <button
            aria-disabled="false"
            class="c9 c10 c11"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
              />
            </svg>
            <span
              class="c2"
            >
              Close modal
            </span>
          </button>
        </header>
        <div
          class="c12 c13"
          dir="ltr"
          style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
        >
          <style>
            [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
          </style>
          <div
            class="c14"
            data-radix-scroll-area-viewport=""
            style="overflow-x: scroll; overflow-y: scroll;"
          >
            <div
              style="min-width: 100%; display: table;"
            >
              <div
                class="c15"
              >
                <div
                  class="c16 c17"
                >
                  <div
                    class="c18"
                  >
                    <label
                      class="c19 c20"
                      for="folder-destination"
                      id="folder-destination-label"
                    >
                      Location
                    </label>
                    <div
                      class=" css-1iyvlxg-container"
                    >
                      <span
                        class="css-1f43avz-a11yText-A11yText"
                        id="react-select-2-live-region"
                      />
                      <span
                        aria-atomic="false"
                        aria-live="polite"
                        aria-relevant="additions text"
                        class="css-1f43avz-a11yText-A11yText"
                        role="log"
                      />
                      <div
                        class=" css-oe0cnm-control"
                      >
                        <div
                          class=" css-onz4xi-ValueContainer"
                        >
                          <div
                            class=" css-1ryi7up-singleValue"
                          >
                            Media Library
                          </div>
                          <input
                            aria-activedescendant=""
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-invalid="false"
                            aria-readonly="true"
                            class="css-mohuvp-dummyInput-DummyInput"
                            id="folder-destination"
                            inputmode="none"
                            role="combobox"
                            tabindex="0"
                            value=""
                          />
                        </div>
                        <div
                          class=" css-1hm8yhz-IndicatorsContainer"
                        >
                          <div
                            aria-hidden="true"
                            class="c21 c22 c23"
                          >
                            <svg
                              fill="currentColor"
                              height="16"
                              viewBox="0 0 32 32"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="m26.708 12.708-10 10a1 1 0 0 1-1.415 0l-10-10A1 1 0 0 1 6 11h20a1 1 0 0 1 .707 1.707"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                      <input
                        name="destination"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="c5 c6 c24"
        >
          <button
            aria-disabled="false"
            class="c25 c26 c27"
            name="cancel"
            type="button"
          >
            <span
              class="c28"
            >
              Cancel
            </span>
          </button>
          <button
            aria-disabled="false"
            class="c25 c26 c29"
            type="submit"
          >
            <span
              class="c28"
            >
              Move
            </span>
          </button>
        </footer>
      </form>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
